### ISSUE TITLE
Web/Desktop - Split bill- Own contact gets highlighted when you press down arrow and the highlight goes out of the screen when press down arrow @Puneet-here

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#10127]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com and log in
2. Go to a Group conversation
3. Click actions > split bill > enter amount > split
4. Press down key

## Expected Result:
Your own contact shouldn't get highlighted as you can't make any interaction with it and when you are on last contact then on pressing down key first contact should be highlighted.

## Actual Result:
Own contact gets highlighted when you press down arrow and the highlight goes out of the screen when press down arrow

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.86.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/181147394-054ba24e-4dfc-49b5-839e-a069bcace6aa.mov

**Expensify/Expensify Issue URL:** 

**Issue reported by:** @Puneet-here

**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1657220128726689

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Puneet-here** [comment]
>highlight goes out of the screen when press down arrow

This issue happens at every list. If the first contact is highlighted then on pressing up arrow key, the highlight goes away and same happens if you are at the last user and press down key.

**Comment by robertjchen** [comment]
Definitely not ideal for a clean experience. Looks like this might only be an issue on Web/Desktop- let's open this up for proposals from the team 🤔

**Comment by NicMendonca** [comment]
job post: https://www.upwork.com/jobs/~01bda8500853223a95

**Comment by syedsaroshfarrukhdot** [comment]
**Propsosal**

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/ArrowKeyFocusManager.js#L36-L38

The bug that causes highlighter to hide  when up arrow on keyboard is pressed at the list top item in all list is because newFocusedIndex is -1 at that time due to this line 

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/ArrowKeyFocusManager.js#L48

The bug that causes highlighter to hide  when down arrow on keyboard is pressed at the list bottom item in all list is because newFocusedIndex > this.props.maxIndex beacuse of this condition incrementing it every time down arrow key is pressed newFocusedIndex incremented and it becomes equal to this.props.maxIndex at which  newFocusedIndex = 0; is setted so highlighter hides and when again arrow key is pressed it higlights the top list item. 

The other bug with own contact getting highlighted is because of the using a genric  ArrowKeyFocusManager on all the lists being displayed at IOUConfirmationList 

**Solution**

To fix bug with up arrow on keyboard hiding highlighter we can do newFocusedIndex = this.props.maxIndex - 1; so when up key is pressed at the top list item instead of being hide it highlights the bottom item

To fix bug with down arrow on keyboard hiding highlighter we can do  if (newFocusedIndex > this.props.maxIndex - 1 ) so when down key is pressed at the bottom list item instead of being hide it highlights the bottom item we set  newFocusedIndex = 0 which highlights the top list item.

To fix the bug with own contact being highlighted we need to send a props which differntiate the split group bill because it is the only view having displayed own email in the list like any other  IOUConfirmationList in the app.

so we could do something like this we could pass props to ArrowKeyFocusManager  isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false} because it is the only component which have section WHO PAID in IOUConfirmationList

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/OptionsSelector.js#L378-L382

 <ArrowKeyFocusManager
                focusedIndex={this.state.focusedIndex}
                maxIndex={this.props.canSelectMultipleOptions ? this.state.allOptions.length : this.state.allOptions.length - 1}
                onFocusedIndexChanged={this.props.disableArrowKeysActions ? () => {} : this.updateFocusedIndex}
                isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false}
  >

In ArrowKeyFocusManager we can handle the highlight index and skip first list item when isSplitBill is true


        this.unsubscribeArrowUpKey = KeyboardShortcut.subscribe(arrowUpConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill

            if (this.props.maxIndex < 1) {
                return;
            }

            let newFocusedIndex = this.props.focusedIndex - 1;

            // Wrap around to the bottom of the list
            if (newFocusedIndex < focusIndexStartIndex ? 1 : 0) {
                newFocusedIndex = this.props.maxIndex - 1;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowUpConfig.descriptionKey, arrowUpConfig.modifiers, true);

        this.unsubscribeArrowDownKey = KeyboardShortcut.subscribe(arrowDownConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill
           


            if (this.props.maxIndex < 1) {
                return;
            }
            let newFocusedIndex = this.props.focusedIndex + 1;

           

            // Wrap around to the top of the list
            if (newFocusedIndex > this.props.maxIndex - 1 ) {
                newFocusedIndex = focusIndexStartIndex ? 1 : 0;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowDownConfig.descriptionKey, arrowDownConfig.modifiers, true);


After Fixing 

https://user-images.githubusercontent.com/81307212/182759852-7d936ea3-496a-4406-bd5f-7a5397c6a51a.mov

**Comment by syedsaroshfarrukhdot** [comment]
**Proposal**

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/ArrowKeyFocusManager.js#L36-L38

The bug that causes highlighter to hide  when up arrow on keyboard is pressed at the list top item in all list is because newFocusedIndex is -1 at that time due to this line 

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/ArrowKeyFocusManager.js#L48

The bug that causes highlighter to hide  when down arrow on keyboard is pressed at the list bottom item in all list is because newFocusedIndex > this.props.maxIndex beacuse of this condition incrementing it every time down arrow key is pressed newFocusedIndex incremented and it becomes equal to this.props.maxIndex at which  newFocusedIndex = 0; is setted so highlighter hides and when again arrow key is pressed it higlights the top list item. 

The other bug with own contact getting highlighted is because of the using a genric  ArrowKeyFocusManager on all the lists being displayed at IOUConfirmationList 

**Solution :**


To fix bug with up arrow on keyboard hiding highlighter we can do newFocusedIndex = this.props.maxIndex - 1; so when up key is pressed at the top list item instead of being hide it highlights the bottom item

To fix bug with down arrow on keyboard hiding highlighter we can do  if (newFocusedIndex > this.props.maxIndex - 1 ) so when down key is pressed at the bottom list item instead of being hide it highlights the bottom item we set  newFocusedIndex = 0 which highlights the top list item.

To fix the bug with own contact being highlighted we need to send a props which differntiate the split group bill because it is the only view having displayed own email in the list like any other  IOUConfirmationList in the app.

so we could do something like this we could pass props to ArrowKeyFocusManager  isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false} because it is the only component which have section WHO PAID in IOUConfirmationList

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/OptionsSelector.js#L378-L382

 We pass this props to ArrowKeyFocusManager isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false}
### <ArrowKeyFocusManager
                focusedIndex={this.state.focusedIndex}
                maxIndex={this.props.canSelectMultipleOptions ? this.state.allOptions.length : this.state.allOptions.length - 1}
                onFocusedIndexChanged={this.props.disableArrowKeysActions ? () => {} : this.updateFocusedIndex}
                isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false}
  >

In ArrowKeyFocusManager we can handle the highlight index and skip first list item when isSplitBill is true


        this.unsubscribeArrowUpKey = KeyboardShortcut.subscribe(arrowUpConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill

            if (this.props.maxIndex < 1) {
                return;
            }

            let newFocusedIndex = this.props.focusedIndex - 1;

            // Wrap around to the bottom of the list
            if (newFocusedIndex < focusIndexStartIndex ? 1 : 0) {
                newFocusedIndex = this.props.maxIndex - 1;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowUpConfig.descriptionKey, arrowUpConfig.modifiers, true);

        this.unsubscribeArrowDownKey = KeyboardShortcut.subscribe(arrowDownConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill
           


            if (this.props.maxIndex < 1) {
                return;
            }
            let newFocusedIndex = this.props.focusedIndex + 1;

           

            // Wrap around to the top of the list
            if (newFocusedIndex > this.props.maxIndex - 1 ) {
                newFocusedIndex = focusIndexStartIndex ? 1 : 0;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowDownConfig.descriptionKey, arrowDownConfig.modifiers, true);


After Fixing 

https://user-images.githubusercontent.com/81307212/182759852-7d936ea3-496a-4406-bd5f-7a5397c6a51a.mov

**Comment by syedsaroshfarrukhdot** [comment]
**Proposal**

The Issue with highlighter being hidden from top when up key is pressed and from bottom when down key is pressed in all lists is due to maxIndex props being passed to ArrowKeyFocusManager when this.props.canSelectMultipleOptions is true it passes the length to it. In case if this.state.allOptions.length is equal to 2 ArrowKeyFocusManager starts from 0 to 2 highlighting item at index 0 and 1 as there is no item at index 2 it hides and when again button is pressed it is resetted.

https://github.com/Expensify/App/blob/f5f076a13bb4a991dbeaf32a8f7a7b0625e8a3cc/src/components/OptionsSelector/BaseOptionsSelector.js#L269

The other bug with own contact getting highlighted is because of the using a genric  ArrowKeyFocusManager on all the lists being displayed at IOUConfirmationList.  So when  own user email is added in the list it also highlights itself

**Solution :**

To fix bug with up arrow and down arrow on keyboard hiding highlighter we can do maxIndex={this.state.allOptions.length - 1}. Doing this fix highlighter issue in all lists.


To stop own contact getting highlighted we could do something like this we could pass props to ArrowKeyFocusManager  isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false} because it is the only component which have section WHO PAID in IOUConfirmationList

https://github.com/Expensify/App/blob/151ec1ad3355cfa162ad490030f19a7134bbc076/src/components/OptionsSelector.js#L378-L382

 We pass this props to ArrowKeyFocusManager isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false}
### <ArrowKeyFocusManager
                focusedIndex={this.state.focusedIndex}
                maxIndex={this.props.canSelectMultipleOptions ? this.state.allOptions.length : this.state.allOptions.length - 1}
                onFocusedIndexChanged={this.props.disableArrowKeysActions ? () => {} : this.updateFocusedIndex}
                isSplitBill={this.props.sections[0].title === "WHO PAID?" ? true : false}
  >

In ArrowKeyFocusManager we can handle the highlight index and skip first list item when isSplitBill is true


        this.unsubscribeArrowUpKey = KeyboardShortcut.subscribe(arrowUpConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill

            if (this.props.maxIndex < 1) {
                return;
            }

            let newFocusedIndex = this.props.focusedIndex - 1;

            // Wrap around to the bottom of the list
            if (newFocusedIndex < focusIndexStartIndex ? 1 : 0) {
                newFocusedIndex = this.props.maxIndex;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowUpConfig.descriptionKey, arrowUpConfig.modifiers, true);

        this.unsubscribeArrowDownKey = KeyboardShortcut.subscribe(arrowDownConfig.shortcutKey, () => {
            let focusIndexStartIndex = this.props.isSplitBill

            if (this.props.maxIndex < 1) {
                return;
            }

            let newFocusedIndex = this.props.focusedIndex + 1;

            // Wrap around to the top of the list
            if (newFocusedIndex > this.props.maxIndex) {
                newFocusedIndex = focusIndexStartIndex ? 1 : 0;
            }

            this.props.onFocusedIndexChanged(newFocusedIndex);
        }, arrowDownConfig.descriptionKey, arrowDownConfig.modifiers, true);

After Fixing 

https://user-images.githubusercontent.com/81307212/182759852-7d936ea3-496a-4406-bd5f-7a5397c6a51a.mov

**Comment by mananjadhav** [comment]
@syedsaroshfarrukhdot I prefer not to add module specific props to `ArrowFocusManager` component. I expect the it be generic (I am working on another PR which will use it on a list of `MenuItems`) . 

Still open for better proposals.
--------------

Also team,
I just ran this on staging.new.expensify.com. I can see the highlight goes back to the first item in the list (v1.1.88-7).  Are we looking to skip the first record (own contact) and make the arrows work only in the `Who was there` section?

https://user-images.githubusercontent.com/3069065/184504547-fb87c57d-f9b2-4c24-b0a9-34ef7be75acf.mp4

@kbecciv Can you confirm if you're able to reproduce the list movement issue?

**Comment by MonilBhavsar** [comment]
Good point. I think since `Who Paid?` option is disabled, it makes sense to just move through `Who was there?`

**Comment by vladnobenladen** [comment]
### **Proposal**
We can introduce new prop called `disabledIndexes` in `ArrowKeyFocusManager` component.
What does this prop do?
**Step 1** When arrow up/down key is pressed, check if `newFocusedIndex` exists in `disabledIndexes` list.
**Step 2** If not exist, do normal flow (call `this.props.onFocusedIndexChanged(newFocusedIndex)`). If exists, go to **Step 3**
**Step 3** Trigger the same key again (if up, should be up; if down, should be down). This should go to **Step 1**
So the idea is to do repeat triggering the same key until enabled option is selected
Please check full code proposal here:
https://github.com/vladnobenladen/Expensify/commit/5692a543ac17bd6e012ea7cf7bb27d38cdc9e44b
This is the key to fix this GH:
<img width="931" alt="proposal" src="https://user-images.githubusercontent.com/97473779/185079423-23dcb179-b9d7-4faa-a6e8-066864af60e5.png">

**Comment by mananjadhav** [comment]
@vladnobenladen Thanks for the proposal. Looks like adding `disabledIndexes` would work but lets take this example if in a list I am adding disabledIndexes in between, then it would skip the ones and what happens if I have 2-3 consecutive disabled Indexes? Can you tell me what is the UI behavior look like to the user?

**Comment by vladnobenladen** [comment]
@mananjadhav Let me take example: List has 5 options with consecutive disabled ones
Option 1
Option 2
Option 3 (disabled)
Option 4 (disabled)
Option 5

- press Down arrow -> Option 1 selected
- press Down arrow again -> Option 2 selected
- press Down arrow again -> Option 5 selected
- press Down arrow again -> Option 1 selected
...

Another example, when last option is disabled:
Option 1
Option 2
Option 3 (disabled)
Option 4 (disabled)
Option 5 (disabled)

- press Down arrow -> Option 1 selected
- press Down arrow again -> Option 2 selected
- press Down arrow again -> Option 1 selected
...

**Comment by mananjadhav** [comment]
I understood what happen in terms of code. What I meant to ask was, in terms of UI do the users see a jump or it directly goes to the n-th item.

Also in terms of implementation, I don't like the idea of unnecessarily add recursion, when this can be covered by simple index handling.

**Comment by vladnobenladen** [comment]
> I understood what happen in terms of code. What I meant to ask was, in terms of UI do the users see a jump or it directly goes to the n-th item.
> 
> Also in terms of implementation, I don't like the idea of unnecessarily add recursion, when this can be covered by simple index handling.

Ah, it directly goes to n-th item of course in UI view. Jump is done in background

**Comment by vladnobenladen** [comment]
> Also in terms of implementation, I don't like the idea of unnecessarily add recursion, when this can be covered by simple index handling.
If it's an issue, I can update proposal to use for or while loop instead of recursive function

**Comment by vladnobenladen** [comment]
@mananjadhav please check https://github.com/vladnobenladen/Expensify/commit/0a220e3fd8f0f2aebb3f3c9796ca5265a147ba03

https://user-images.githubusercontent.com/97473779/185301760-83f3d02d-061c-474a-8021-61374103e769.mov

**Comment by mananjadhav** [comment]
> @mananjadhav please check [vladnobenladen@0a220e3](https://github.com/vladnobenladen/Expensify/commit/0a220e3fd8f0f2aebb3f3c9796ca5265a147ba03)

@MonilBhavsar, @vladnobenladen's updated proposal here looks good to me. 

🎀 👀 🎀 
C+ reviewed

**Comment by MonilBhavsar** [comment]
Looks good to me too. Thanks!

**Comment by NicMendonca** [comment]
No overdue 👌 Latest update on the PR [here ](https://github.com/Expensify/App/pull/10477#issuecomment-1232417939)

**Comment by MonilBhavsar** [comment]
PR was merged

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by NicMendonca** [comment]
@0xaim don't forget to apply to the [Upwork job ](https://www.upwork.com/jobs/~01bda8500853223a95)so we can issue payment. Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by NicMendonca** [comment]
@Puneet-here don't forget to accept the Upwork offer so I can issue payment. Thanks!

**Comment by NicMendonca** [comment]
@mananjadhav paid - thank you!

**Comment by Puneet-here** [comment]
Thanks for the ping. I have accepted the offer now.

**Comment by NicMendonca** [comment]
@Puneet-here - paid, thank you!

**Comment by MonilBhavsar** [comment]
@NicMendonca can we close this if we have issued payment to everyone involved?

**Comment by NicMendonca** [comment]
@MonilBhavsar we haven't! Still waiting on @0xaim to apply for the job so that I can issue payment. 

Bump @0xaim! Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by NicMendonca** [comment]
@0xaim sure! Just ping here once you've applied for the job. Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fourth week)

**Comment by NicMendonca** [comment]
@0xmiroslav thank you! Just need you to accept the offer as a last step here.

**Comment by mananjadhav** [comment]
@NicMendonca Are we done with the payment and good to close this?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fifth week) Are we ok to close?

**Comment by NicMendonca** [comment]
@0xmiroslav paid- thank you!
